recirq/optimize/mpg.py

```diff
@@ -0,0 +1,390 @@
+    def __init__(self, learning_rate, decay_steps, decay_rate, staircase=False):
+        """The Exponential schedule for some hyperparameter (e.g. learning_rate)
+        
+        Exponential decay for the `learning rate`. For each `decay_steps`, the learning 
+        rate is scheduled to decay at the `decay_rate`. The `staircase` controls whether 
+        to decay smoothly or discontinuously. After this many timesteps pass, the final
+        learning rate is returned.
+
+        Args: 
+            learning rate: the initial learning rate 
+            decay_steps: the learning rate is scheduled to decay every such number of steps 
+            decay_rate: the learning rate is scheduled to decay at the such rate
+            staircase: if True, the learning rate keeps the same before every decay steps; 
+                       otherwise, the learning rate decays smoothly according 
+                       to exponential interpolation.
+        
+        Returns: 
+            a class of the schedule
+        """
+        self.learning_rate = learning_rate
+        self.decay_steps = decay_steps
+        self.decay_rate = decay_rate
+        self.staircase = staircase
+
+    def value(self, t):
+        """Return the value of the schedule at time step t 
+
+        Args: 
+            t: the time step for the schedule 
+
+        Returns: 
+            the schedule value
+        """
+        m = t / self.decay_steps
+        if self.staircase == True:
+            m = np.floor(m)
+
+        return self.learning_rate * self.decay_rate ** m
+
+
+class Adam:
+    def __init__(
+        self, lr_schedule=ConstantSchedule(0.001), b1=0.9, b2=0.999, eps=10 ** -8
+    ):
+        """The optimizer Adam: a method for stochastic gradient descent  
+
+        Adam as described in http://arxiv.org/pdf/1412.6980.pdf.
+        adapted from https://github.com/HIPS/autograd/blob/master/autograd/misc/optimizers.py#L57-L71
+
+        Args:
+            lr_schedule: the class of learning rate decay schedule. Defaults to ConstantSchedule(0.001).
+            b1 (float, optional): coefficients used for computing
+                running averages of gradient. Defaults to 0.9.
+            b2 (float, optional): coefficients used for computing
+                running averages of gradient's square. Defaults to 0.999.
+            eps (float, optional): term added to the denominator to improve
+                numerical stability. Defaults to 10**-8.
+        """
+        self.b1 = b1
+        self.b2 = b2
+        self.eps = eps
+        self.step = 0
+        self.lr_schedule = lr_schedule
+
+    def ascend(self, grad, x):
+        """Performs a single optimization step.
+
+
+        Args:
+            grad: the gradient computed at the current update.  
+            x: the current value of the parameters. 
+
+        Returns:
+            the updated parameter values. 
+        """
+        if self.step == 0:
+            self.m = np.zeros(len(x))
```

Comment:
you shouldn't initialize class members outside of `__init__`

Reply:
agreed!